DOI.js

```diff
@@ -93,33 +95,47 @@ function completeDOIs(doc) {
 	}
 }
 
-function retrieveDOIs(DOIs, doc) {
-	__num_DOIs = DOIs.length;
+function retrieveDOIs(dois, doc, providers) {
+	var numDois = dois.length;
+	var provider = providers.shift();
+	
+	var remainingDOIs = dois.slice();//copy array but not by reference
 
-	for(var i=0, n=DOIs.length; i<n; i++) {
+	for(var i=0, n=dois.length; i<n; i++) {
 		(function(doc, DOI) {
 			var translate = Zotero.loadTranslator("search");
-			translate.setTranslator("11645bd1-0420-45c1-badb-53fb41eeb753");
+			translate.setTranslator(provider.id);
 	
 			var item = {"itemType":"journalArticle", "DOI":DOI};
 			translate.setSearch(item);
 	
 			// don't save when item is done
 			translate.setHandler("itemDone", function(translate, item) {
+				selectArray[item.DOI] = item.title;
 				if (!item.title) {
-					Zotero.debug("No title available for " + DOI);
-					return;
+					Zotero.debug("No title available for " + item.DOI);
+					item.title = "[No Title]";
+					selectArray[item.DOI] = "[" + item.DOI + "]";
+				}
+				items[item.DOI] = item;
+
+				// done means not remaining anymore
+				if (remainingDOIs.indexOf(item.DOI) > -1) {
+					remainingDOIs.splice(remainingDOIs.indexOf(item.DOI), 1);
+				} else {
+					Z.debug(item.DOI + " not anymore in the list of remainingDOIs = " + remainingDOIs);
```

Comment:
Thanks @avram . Yes, it works. Here is the debug output if I copy the line also to the `if` statement:
```
23:51:58 [
             "0": "10.1057/9780230391116"
             "1": "10.1073/pnas.1212247109"
             "2": "10.3205/mbi000337"
             "3": "10.3205/mbi000336"
             "4": "10.3205/mbi000354"
         ]
23:51:59 10.1057/9780230391116 not anymore in the list of remainingDOIs = 10.1057/9780230391116,10.1073/pnas.1212247109,10.3205/mbi000337,10.3205/mbi000336,10.3205/mbi000354
23:51:59 10.1073/pnas.1212247109 not anymore in the list of remainingDOIs = 10.1073/pnas.1212247109,10.3205/mbi000337,10.3205/mbi000336,10.3205/mbi000354
23:51:59 Done with CrossRef. Remaining DOIs: 10.3205/mbi000337,10.3205/mbi000336,10.3205/mbi000354
23:51:59 10.3205/mbi000336 not anymore in the list of remainingDOIs = 10.3205/mbi000337,10.3205/mbi000336,10.3205/mbi000354
23:51:59 10.3205/mbi000337 not anymore in the list of remainingDOIs = 10.3205/mbi000337,10.3205/mbi000354
23:51:59 10.3205/mbi000354 not anymore in the list of remainingDOIs = 10.3205/mbi000354
23:51:59 Done with DataCite. Remaining DOIs: 
23:52:07 Translation successful
```